packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js

```diff
@@ -55,49 +36,69 @@ export const useMagentoRoute = props => {
         };
     }, []);
 
-    // If the entry for this pathname is a redirect, perform the redirect.
+    const storeCodeResult = useQuery(getStoreCode, {
+        fetchPolicy: 'cache-and-network',
+        nextFetchPolicy: 'cache-first'
+    });
+
+    const { data: storeCodeData } = storeCodeResult;
+    const store = storeCodeData && storeCodeData.storeConfig.code;
+    const routeKey = getRouteKey(pathname, store);
+    const routeData = responses.get(routeKey);
+
+    const urlResult = useQuery(RESOLVE_URL, {
+        fetchPolicy: 'cache-and-network',
+        nextFetchPolicy: 'cache-first',
+        variables: { url: pathname }
+    });
+
+    // fetch a component if necessary
+    useEffect(() => {
+        const { data, error, loading } = urlResult;
+
+        // exit if we've unmounted, or if we don't have data
+        if (!isMountedRef.current || loading || !data || !store) return;
+
+        // the result may be null or incomplete, amounting to a 404
+        const { urlResolver } = data;
+        const { id, redirectCode, relative_url, type } = urlResolver || {};
+        const isNotFound = !urlResolver || !id || !type || id === -1;
+        const isRedirect = REDIRECT_CODES.has(redirectCode);
+
+        getRouteComponent(type, id)
+            .then(component => {
+                // throw Apollo errors to the catch block
+                if (error) throw error;
```

Comment:
So we always want to call `getRouteComponent` even if urlResolver returns an error or `null`?

Reply:
Refactored to avoid extra calls. Seems a lot cleaner, and visibly faster.